src/peft/tuners/lora/model.py

```diff
@@ -431,12 +438,20 @@ def _unload_and_optionally_merge(self, merge=True, progressbar: bool = False, sa
                         new_module = torch.nn.Linear(target.in_features, target.out_features, bias=bias)
                 if merge:
                     target.merge(safe_merge=safe_merge)
-                self._replace_module(parent, target_name, new_module, target)
+
+                if hasattr(module, "_hf_hook") and isinstance(module._hf_hook, AlignDevicesHook):
+                    module._hf_hook.post_forward(module, torch.tensor([]))
+                else:
+                    self._replace_module(parent, target_name, new_module, target)
```

Comment:
Is it correct to put this in the `else` clause? Should it not always be called?

Reply:
The `_replace_module` function unfortunately leads to a memory leak in this case, due to copied tensors of parameters (via `.to(device)`).  This manifests when, say, a model with 11GB of parameters on a 15GB GPU calls this function and experiences OOM.  As this function is mostly a device assignment utility, we may not need it if the device alignment hooks are present.